tests/Microsoft.Bot.Builder.Tests/TurnContextTests.cs

```diff
@@ -291,6 +292,39 @@ void ValidateUpdate(Activity activity)
             Assert.IsTrue(updateResult.Id == "test");
         }
 
+        [TestMethod]
+        public async Task UpdateActivityWithBotFrameworkAdapter()
+        {
+            const string ACTIVITY_ID = "activity ID";
+            const string CONVERSATION_ID = "conversation ID";
+
+            var foundActivity = false;
+
+            var a = new BotFrameworkAdapter(new SimpleCredentialProvider());
```

Comment:
Can we base these tests on TestAdapter or some custom subclass of BotAdapter...

Reply:
Hi @johnataylor 

We certainly can use `TestAdapter` or some custom subclass of `BotAdapter`, but I'm not sure how that would help. There is already a `TurnContext.UpdateActivityAsync` test that uses `SimpleAdapter`. The reason we didn't catch this bug in the first place is because our turn context tests don't use `BotFrameworkAdapter` and so they don't end up calling the conversations operations. You may notice that I'm using a custom `HttpClient` so that no actual requests are made. With that in mind, how would you like me to test to make sure [`Conversations.UpdateActivityAsync`](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Connector/ConversationsEx.cs#L184) doesn't throw an exception when `TurnContext.UpdateActivityAsync` is called with an activity created by `MessageFactory`?

Reply:
The point is that the model is TurnContext adds the address and then calls a BotAdapter. You could use BotFrameworkAdapter as your BotAdapter implementation but then you are adding other hooks to make that story work. It is a more direct and more accurate test to test for exactly what you want which is TurnContext adding the address. (For example supposed BotFrameworkAdapter had been adding the address - the bug here would have been hidden - until someone used a non-BotFrameworkAdapter adapter.

The reason we didn't catch this bug before was because we lacked test coverage on the TurnContext. Its more precisely that that is missing - so I'm saying your tests should fill that gap.

Make sense?

Reply:
Hi @johnataylor - I have responded to you in Teams

Reply:
I've just pushed a fix